### PR TITLE
[SMTChecker] Support addmod and mulmod

### PR DESCRIPTION
**Pull request by leonardoalt**
Ref https://github.com/ethereum/solidity/issues/9908

This PR supports the logic for `addmod` and `mulmod`, and adds verification targets for `k != 0` for both BMC and CHC.
The error message `CHC: Error trying to invoke` happens because Spacer at the moment complains about every `mod`. This will get better after https://github.com/ethereum/solidity/issues/9908. That's also the reason why the newly added error id for `Div by zero` in CHC is not used by the tests and was added as "ignore" in the errors script.

### CODE REVIEW

File: libsolidity/formal/SMTEncoder.cpp [hunk 1]
```diff
@@ -738,6 +742,32 @@ void SMTEncoder::visitGasLeft(FunctionCall const& _funCall)
...
+	auto x = expr(*args.at(0));
+	auto y = expr(*args.at(1));
+	auto k = expr(*args.at(2));
+	m_context.addAssertion(k != 0);
+	defineExpr(_funCall, (x + y) % k);
+}
+
+void SMTEncoder::visitMulMod(FunctionCall const& _funCall)
+{
+	FunctionType const& funType = dynamic_cast<FunctionType const&>(*_funCall.expression().annotation().type);
```
Comment (mijovic) [inline libsolidity/formal/SMTEncoder.cpp]: I think this is great example where to use auto. Even clang-tidy complains about it :)
```suggestion
	auto const& funType = dynamic_cast<FunctionType const&>(*_funCall.expression().annotation().type);
```

File: libsolidity/formal/SMTEncoder.cpp [hunk 2]
```diff
@@ -738,6 +742,32 @@ void SMTEncoder::visitGasLeft(FunctionCall const& _funCall)
 		m_context.addAssertion(symbolicVar->currentValue() <= symbolicVar->valueAtIndex(index - 1));
 }
 
+void SMTEncoder::visitAddMod(FunctionCall const& _funCall)
+{
+	FunctionType const& funType = dynamic_cast<FunctionType const&>(*_funCall.expression().annotation().type);
```
Comment (mijovic) [inline libsolidity/formal/SMTEncoder.cpp]:
```suggestion
	auto const& funType = dynamic_cast<FunctionType const&>(*_funCall.expression().annotation().type);
```

**Comment by leonardoalt** [comment]
Will update this PR after the other one because they conflict

File: libsolidity/formal/SMTEncoder.cpp [hunk 3]
```diff
@@ -738,6 +742,32 @@ void SMTEncoder::visitGasLeft(FunctionCall const& _funCall)
 		m_context.addAssertion(symbolicVar->currentValue() <= symbolicVar->valueAtIndex(index - 1));
 }
 
+void SMTEncoder::visitAddMod(FunctionCall const& _funCall)
```
Comment (mijovic) [inline libsolidity/formal/SMTEncoder.cpp]: Can `visitAddMod` and `visitMulMod` be one function? Code looks the same except last line which differs only by applying `+` or `*` depending on which operation we are in.

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp]: Yea I'll merge them

**Comment by leonardoalt** [comment]
Updated

**Review by hrkrshnn — COMMENTED** [review]
```solidity
contract C {
	function test_addmod(uint x, uint y) public pure {
		require(x % 13 == 0);
		require(y % 13 == 0);

		uint z = addmod(x, y, 13);
		assert(z == 0);
	}
	function test_mulmod(uint x, uint y) public pure {
		require(x % 13 == 0);
		require(y % 13 == 0);

		uint z = mulmod(x, y, 13);
		assert(z == 0);
	}
}
```

Will this test be relevant?

**Comment by leonardoalt** [comment]
@hrkrshnn I can add it, but why `require` instead of `assert`?

**Comment by hrkrshnn** [comment]
@leonardoalt The last one should be assert. Was a typo.

**Comment by mijovic** [comment]
> ```solidity
> contract C {
> 	function test_addmod(uint x, uint y) public pure {
> 		require(x % 13 == 0);
> 		require(y % 13 == 0);
> 
> 		uint z = addmod(x, y, 13);
> 		require(z == 0);
> 	}
> 	function test_mulmod(uint x, uint y) public pure {
> 		require(x % 13 == 0);
> 		require(y % 13 == 0);
> 
> 		uint z = mulmod(x, y, 13);
> 		require(z == 0);
> 	}
> }
> ```
> 
> Will this test be relevant?

You were probably wanted to add assertions at the end of functions, I guess. Nice tests, I think @leonardoalt can add them.

**Review by mijovic — DISMISSED** [review]
Looks good.
If you can add tests @hrkrshnn mentioned would be great

**Comment by leonardoalt** [comment]
Added the test. Unfortunately it's too hard for the solvers =/

File: libsolidity/formal/SMTEncoder.cpp [hunk 4]
```diff
@@ -738,6 +740,24 @@ void SMTEncoder::visitGasLeft(FunctionCall const& _funCall)
...
+	solAssert(kind == FunctionType::Kind::AddMod || kind == FunctionType::Kind::MulMod, "");
+	auto const& args = _funCall.arguments();
+	solAssert(args.at(0) && args.at(1) && args.at(2), "");
+	auto x = expr(*args.at(0));
+	auto y = expr(*args.at(1));
+	auto k = expr(*args.at(2));
+	m_context.addAssertion(k != 0);
+
+	if (kind == FunctionType::Kind::AddMod)
+		defineExpr(_funCall, (x + y) % k);
```
Comment (hrkrshnn) [inline libsolidity/formal/SMTEncoder.cpp:756]: What would be the 'type' of this addition?

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:756]: Integer

**Review by mijovic — DISMISSED** [review]

**Comment by leonardoalt** [comment]
Rebased
@mijovic can you reapprove?

**Review by mijovic — APPROVED** [review]

**Pull request merged**